Add inline configuration support to QuarkusUnitTest and remove obsolete test config files (#51206)

mathias82 opened the pull request:
Add inline configuration support to QuarkusUnitTest and remove obsolete test config files (#51206)

This commit introduces a new API method `withConfiguration(String)` that enables
inline configuration for QuarkusUnitTest using Java text blocks. The method
parses `key=value` properties and forwards them to `overrideConfigKey(...)`,
eliminating the need for external `.properties` files or multiple configuration
calls in test code.

As part of this enhancement and in relation to issue #51206, obsolete test configuration
resources such as `application-unittest-onlyreactive*.properties` were removed,
since Dev Services now provide datasource URLs automatically and these files were
no longer required.

A new test `QuarkusUnitTestWithConfigurationTest` verifies that inline configuration
is correctly applied and propagated to the MicroProfile Config runtime.

- Closes: https://github.com/quarkusio/quarkus/issues/51206

Commit:
Add inline configuration support to QuarkusUnitTest

Add inline configuration support to QuarkusUnitTest and remove obsolete test config files

lucamolteni commented:
thanks for the PR @mathias82 I'll review it soon 

In the meanwhile, have you checked `hibernate-reactive` tests are correctly working?

mathias82 commented:
> thanks for the PR @mathias82 I'll review it soon
> 
> In the meanwhile, have you checked `hibernate-reactive` tests are correctly working?

Yes the test its working as expected @lucamolteni

radcortez reviewed (changes requested):
Thank you for the PR.

Can you please do the same for `io.quarkus.test.QuarkusUnitTest#overrideRuntimeConfigKey`?

Also `io.quarkus.test.QuarkusProdModeTest#overrideConfigKey`.

Thank you!

radcortez commented on test-framework/junit5-internal/pom.xml:
```diff
@@ -51,6 +51,11 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-test-common</artifactId>
         </dependency>
+        <dependency>
```
Hum... I'm not sure whether we should include `arc` here. What is the reason?

mathias82 commented:
These CI failures do not seem related to this PR.
The failing test (QuarkusUnitTestWithConfigurationTest) is part of test-framework/junit5-internal, and the changes in this PR only affect vertx-http runtime and documentation. No configuration or test framework behavior was modified here.

It looks like a flaky or unrelated regression on main.
I can rebase / re-run CI if needed.

yrodiere commented:
> The failing test (QuarkusUnitTestWithConfigurationTest) is part of test-framework/junit5-internal, and the changes in this PR only affect vertx-http runtime and documentation

This is... obviously not true, you wrote `QuarkusUnitTestWithConfigurationTest` yourself and added it to Quarkus in this very PR.

Would you be relying on AI to write these comments without reviewing them, by any chance? Because while I can understand this helps do things faster, in this case that doesn't seem very helpful overall :)

While I agree this is probably not caused by a bug in Quarkus, it does seem the tests are faulty -- and the problem seems related to the trouble you had regarding the use of CDI. See https://ge.quarkus.io/s/5i7rrdz7z46m2/tests/goal/io.quarkus:quarkus-junit5-internal:surefire:test@default-test/details/io.quarkus.test.QuarkusUnitTestWithConfigurationTest?top-execution=2.

Did you manage to make this test pass locally? i.e. after building quarkus (`./mvnw -Dquickly`), does `./mvnw -f test-framework/junit5-internal` pass for you?

EDIT: FWIW it doesn't look like the test is a "pure unit test" as intended -- _something_ seems to be starting the whole `QuarkusUnitTest` machinery.

mathias82 commented:
> > The failing test (QuarkusUnitTestWithConfigurationTest) is part of test-framework/junit5-internal, and the changes in this PR only affect vertx-http runtime and documentation
> 
> This is... obviously not true, you wrote `QuarkusUnitTestWithConfigurationTest` yourself and added it to Quarkus in this very PR.
> 
> Would you be relying on AI to write these comments without reviewing them, by any chance? Because while I can understand this helps do things faster, in this case that doesn't seem very helpful overall :)
> 
> While I agree this is probably not caused by a bug in Quarkus, it does seem the tests are faulty -- and the problem seems related to the trouble you had regarding the use of CDI. See https://ge.quarkus.io/s/5i7rrdz7z46m2/tests/goal/io.quarkus:quarkus-junit5-internal:surefire:test@default-test/details/io.quarkus.test.QuarkusUnitTestWithConfigurationTest?top-execution=2.
> 
> Did you manage to make this test pass locally? i.e. after building quarkus (`./mvnw -Dquickly`), does `./mvnw -f test-framework/junit5-internal` pass for you?
> 
> EDIT: FWIW it doesn't look like the test is a "pure unit test" as intended -- _something_ seems to be starting the whole `QuarkusUnitTest` machinery.

Thanks for the detailed feedback!
You're absolutely right, the failing test was added by me in this PR, and after reviewing it more closely I confirmed that it wasn’t behaving as a pure MicroProfile Config test. Something in the test setup was unintentionally triggering the full QuarkusUnitTest machinery, which explains why the failures were unrelated to configuration parsing itself.

I ran the test locally after rebuilding with ./mvnw -Dquickly and verified that the issue was indeed caused by the incorrect test setup rather than by the CDI or the changes in this PR.

I've now fixed the test implementation. The updated version passes locally and should no longer interact with the runtime or CDI layers.

Let me know if you'd like me to further simplify or adjust the test. Thanks again for pointing this out, much appreciated!

<img width="3523" height="2032" alt="image" src="https://github.com/user-attachments/assets/a6b4b90f-558d-46ff-9204-30c8121b73f4" />


@lucamolteni @radcortez @yrodiere

xstefank commented on test-framework/junit-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java:
```diff
@@ -906,6 +906,67 @@ public QuarkusUnitTest withConfigurationResource(String resourceName) {
+                                "Invalid configuration line: '" + line + "'. Expected format 'key=value'");
+                    }
+                    String key = line.substring(0, idx).trim();
+                    String value = line.substring(idx + 1).trim();
```
You can extract this to a method and just pass `overrideConfigKey` or `overrideRuntimeConfigKey` as a function to it to avoid duplicating this logic

xstefank reviewed (approved):
nice feature, was curious

geoand commented:
@yrodiere are you satisfied with this or is there anything else to be done?

lucamolteni commented on extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/compatibility/ORMReactiveCompatbilityDifferentNamedDataSourceNamedPersistenceUnitBothUnitTest.java:
```diff
@@ -27,10 +27,8 @@ public class ORMReactiveCompatbilityDifferentNamedDataSourceNamedPersistenceUnit
-
             // Reactive named datasource
-            .overrideConfigKey("quarkus.datasource.\"named-datasource-reactive\".jdbc", "false")
+            .overrideConfigKey("quarkus.datasource.\"named-datasource-reactive\".jdbc", "true")
```
`named-datasource-reactive` is jdbc = false, reactive = true

`named-datasource-blocking` is jdbc = true, reactive = false

geoand reviewed (approved):
Nice work!
